### DISCUSSION TITLE
Apache Kafka Connect single message transforms

### DISCUSSION THREAD

**Discussion by gashcrumb** [atlasmap/atlasmap#3231] [Design & Features]
The Kafka connect API contains among other things an [interface for doing transformations on messages](https://www.confluent.io/blog/kafka-connect-single-message-transformation-tutorial-with-examples/) as they flow from sources to sinks.  It'd be interesting to see if we could leverage these interfaces to allow users to define these transformations in AtlasMap.

**Comment by igarashitm** [comment]
Glueing data shapes for this usecase is something we have to consider well. There could be 2 options:
### Serialized JSON/XML/CSV as a single key/value
In this case kafka schema is not involved and instead the schema has to be fed into AtlasMap UI individually. For example, using `FileStreamSource` connector, a record is sent for each line in the input file, having the text data as a value with null key. Although this case leverages AtlasMap's cross format data mapping feature, it doesn't look like a common usage for kafka.
##### AtlasMap UI enhancements
* Just a regular schema import we have today in AtlasMap, no enhancement needed.
##### AtlasMap backend/runtime enhancements
* #3232 - A small SMT that reads the serialized Document from the key or value of kafka record, pass it to AtlasMap session, and put the result into the key or value of kafka record. No need to maintain kafka schema as it's kafka unaware.
##### Beyond the 1st step
* How/What to integrate the AtlasMap UI into? where to import the schema from, where to export ADM file, how to generate the Kafka Connect configuration (ADM file path, key/value flag) that enables AtlasMap SMT, in some Kafka Connect enabled UI.
* We might want to use both of key and value, like serialized JSON as a key, and serialized XML as a value. Then it also has to retrieve generated Document ID when it's imported into AtlasMap and put them into AtlasMap SMT configuration.
  * Document ID could be fed from outside like we do in Syndesis, where it's guaranteed not to conflict at host side

### kafka structured data as a key and a value
The kafka connect can set up a converter to serialize/deserialize from/to kafka internal data structure. We'll need to create a kafka AtlasMap module and let it handle the mapping between kafka data structure and AtlasMap one. It would be simpler/smaller than others like JSON/XML since kafka's is very similar to what AtlasMap do with AtlasMap `Field`. e.g. The kafka `Struct` is an equivalent to `FieldType.COMPLEX` in AtlasMap. 
How the schema is passed into AtlasMap is a question, especially in AtlasMap UI. It is recommended to use [apicurio-registry](https://github.com/apicurio/apicurio-registry) to manage kafka schema. Putting integration with the schema-registry aside, for the 1st attempt, we would assume the user would manually load the schema into AtlasMap UI just like we expect in standalone usage today. But even so, we'd have to support Avro schema and protobuf schema in AtlasMap.
Also, instead of deliverying kafka record as a single AtlasMap `Document`, key and value are better to be handled as a separate `Document` as they have separate kafka runtime schema maintained for each, attached to the record. Otherwise we'd have to introduce some grouping feature in AtlasMap UI to group key and value with separate schemas. We'd need to investigate how we can attach those kafka runtime schema as a generic runtime `Document` metadata into AtlasMap context.
Or kafka schema might not really required for AtlasMap runtime? in theory all the schemas are already fed into AtlasMap UI at design time, so at runtime there shouldn't be a need for updating schema but the design time schemas could be used as-is for both of source and target.
In this usecase, AtlasMap has to internally use `KafkaField` that handles kafka data structure like `Struct`. AtlasMap never knows the real data format, e.g. JSON, CSV, XML - seralize/desrialize is done by kafka connect. That means plain Avro support in AtlasMap if there is, would be a different implementation than kafka Avro support.
##### AtlasMap UI enhancements
* #3234 - Add new import option for kafka aware schemas (Avro/Protobuf/JSON) and use kafka inspection for that.
##### AtlasMap backend/runtime enhancements
* #3233 - Add kafka module
  * kafka inspection supports Avro schema, Protobuf schema and JSON schema. Wondering if we could reuse a routine from Kafka Connect or apicurio-registry to parse those schemas and build a tree of kafka data structure, which should be easy to convert into AtlasMap inspection result.
  * kafka field reader/writer handles kafka data structure like `Struct`
  * A SMT that reads the Document from the key or value of kafka record, pass it to AtlasMap session, and put the result into the key and value of kafka record. Also put the kafka schema into a target kafka record.
##### Beyond the 1st step
* How/What to integrate the AtlasMap UI into? where to import the schema from (integrate with apicurio-registry?), where to export ADM file, how to generate the Kafka Connect configuration (ADM file path, key/value flag, schema) that enables AtlasMap SMT, in some Kafka Connect enabled UI.
* We might want to use both of key and value. Then it also has to retrieve generated Document ID when it's imported into AtlasMap and put them into AtlasMap SMT configuration with corresponding schema.
  * Document ID could be fed from outside like we do in Syndesis, where it's guaranteed not to conflict at host side

Also note that the 2 options above could be mixed, e.g. read a serialized JSON and map into kafka data structure with Avro schema provided, and vise versa. So the AtlasMap SMT would be a single component and  specify which option is used in SMT configuration.
